File: courseSchedule.go
// Time Complexity : O(v+e)
// Space Complexity : O(v+e)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :

// Your code here along with comments explaining your approach
/*
create a vert array with increment of all dependecies
create map from independent to dependent edges
create q and add all independent vertices
while q is not empty
pop from q, get all dependents from map
for each dependent reduce count in vert array for that element
add all elements with value 0 into q
after loop is finished check if vert has any value greater than 0 if so return false else return true
*/
package main

import "fmt"

func canFinish(numCourses int, prerequisites [][]int) bool {
	vert := make([]int, numCourses)
	m := make(map[int][]int)

	for i:=0;i<len(prerequisites);i++ {
		vert[prerequisites[i][0]]++
	}
	for i:=0;i<len(prerequisites);i++ {
		if _, e := m[prerequisites[i][1]]; e {
			m[prerequisites[i][1]] = append(m[prerequisites[i][1]], prerequisites[i][0])
		} else {
			m[prerequisites[i][1]] = append([]int{}, prerequisites[i][0])
		}
	}

	q := make([]int, 0)
	for i := 0; i < numCourses; i++ {
		if vert[i] == 0 {
			q = append(q, i)
		}
	}

	for len(q) != 0 {
		course := q[0]
		subClasses := m[course]
		for i := 0; i < len(subClasses); i++ {
			vert[subClasses[i]]--
			if vert[subClasses[i]] == 0 {
				q = append(q, subClasses[i])
			}
		}
		q = q[1:]
	}

	for i := 0; i < numCourses; i++ {
		if vert[i] != 0 {
			return false
		}
	}

	return true
}

func MainCourseSchedule() {
	fmt.Println(canFinish(2, [][]int{[]int{1,0}})) //expected true
}

File: course-schedule.js
// Time Complexity : O(V + E)
// Space Complexity : O(V + E)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Lots of variable names, so it took a while just to get it to work.
// Other than that, no problem.

/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {boolean}
 */
var canFinish = function(numCourses, prerequisites) {
    if (numCourses == null || numCourses == 0) return true;
    if (prerequisites == null || prerequisites.length == 0) return true;
    
    let degrees = new Array(numCourses).fill(0),
        map = new Map,
        queue = [];
    for (let i = 0; i < prerequisites.length; i++) {
        degrees[prerequisites[i][0]]++;
        if (!map.has(prerequisites[i][1])) {
            map.set(prerequisites[i][1], []);
        }
        map.get(prerequisites[i][1]).push(prerequisites[i][0]);
    }
    for (let i = 0; i < degrees.length; i++) {
        if (degrees[i] == 0) queue.push(i);
    }
    if (queue.length == 0) return false;
    while (queue.length > 0) {
        let course = queue.shift();
        let children = map.get(course);
        if (children) {
            for (const child of children) {
                degrees[child]--;
                if (degrees[child] == 0) {
                    queue.push(child);
                }
            }
        }
    }
    for (let i = 0; i < degrees.length; i++) {
        if (degrees[i] > 0) return false;
    }
    return true;
};
